text/0000-read-all.md

```diff
@@ -0,0 +1,99 @@
+- Feature Name: read_exact and read_full
+- Start Date: 2015-03-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Rust's `Write` trait has `write_all`, which is a convenience method that calls
+`write` repeatedly to write an entire buffer. This proposal adds two similar
+convenience methods to the `Read` trait: `read_full` and `read_exact`.
+`read_full` calls `read` repeatedly until the buffer has been filled, EOF has
+been reached, or an error other than `Interrupted` occurs. `read_exact` is
+similar to `read_full`, except that reaching EOF before filling the buffer is
+considered an error.
+
+# Motivation
+
+The `read` method may return fewer bytes than requested, and may fail with an
+`Interrupted` error if a signal is received during the call. This requires
+programs wishing to fill a buffer to call `read` repeatedly in a loop. This is
+a very common need, and it would be nice if this functionality were provided in
+the standard library. Many C and Rust programs have the same need, and solve it
+in the same way. For example, Git has [`read_in_full`][git], which behaves like
+the proposed `read_full`, and the Rust byteorder crate has
+[`read_full`][byteorder], which behaves like the proposed `read_exact`.
+[git]: https://github.com/git/git/blob/16da57c7c6c1fe92b32645202dd19657a89dd67d/wrapper.c#L246
+[byteorder]: https://github.com/BurntSushi/byteorder/blob/2358ace61332e59f596c9006e1344c97295fdf72/src/new.rs#L184
+
+# Detailed design
+
+The following methods will be added to the `Read` trait:
+
+``` rust
+fn read_full(&mut self, buf: &mut [u8]) -> Result<usize>;
```

Comment:
@alexcrichton 

> there's a lot of overlap between the two

Yes, that is why `read_exact` is proposed as a simple wrapper around `read_full` (this has other advantages, as discussed above). `read_full` is the general convenience method to handle filling a buffer. `read_exact` just adds a small amount of convenience by converting an EOF to an error. `read_full` is useful any time you want to read a certain amount of data, but an EOF isn't an error. One example would be implementing something like `dd` where one wants to process a file in fixed sized blocks. Another would be working with a protocol that transmitted data in fixed-size chunks, but where the last chunk can be short if the data isn't a multiple of the chunk payload size. It's also matches the `read_in_full` function that git uses internally.

`read_full` does the almost all of the work, making it very simple to perform the equivalent of `read_exact` using `read_full`. Thus, if only one of these functions were to be included, I would prefer it to be `read_full` as it is more general and covers the majority of the convenience.